packages/babel-parser/src/parser/expression.js

```diff
@@ -641,7 +645,11 @@ export default class ExpressionParser extends LValParser {
         this.finishOptionalCallExpression(node);
       }
 
-      if (possibleAsync && this.shouldParseAsyncArrow()) {
+      if (
+        maybeAsyncArrow &&
+        !this.canInsertSemicolon() &&
```

Comment:
`!this.canInsertSemicolon()` prevents 
```js
async (x) 
=> {}
```

Reply:
Maybe we could move the check inside `shouldParseAsyncArrow`?
Also note that both flow and TS allow this code:
```js
async (f)
: t => { }
```
but disallow

```js
async (f) : t
=> { }
```

Reply:
Thanks, I added testcases and moved it into `shouldParseAsyncArrow `